File: updater/src/main/java/updater/identifiers/Node.java
package updater.identifiers;

import org.objectweb.asm.Type;
import updater.utility.AbstractIdentifier;
import updater.utility.ClassWrapper;
import updater.utility.TypeUtilities;

public class Node extends AbstractIdentifier {

    @Override
    public boolean isMatch(ClassWrapper classNode) {
        return classNode.getInstanceFields().size() == 3 &&
                classNode.getCountFieldsOfType(Type.LONG_TYPE) == 1 &&
                classNode.getCountFieldsOfType(Type.getObjectType(classNode.getName())) == 2 &&
                classNode.isOwnerless();
    }

    public MethodIdentifier hasNext = methodIdentifier(methodWrapper -> methodWrapper.returnTypeEquals(Type.BOOLEAN_TYPE));

    public FieldIdentifier key = fieldIdentifier(fieldWrapper -> fieldWrapper.isOfType(Type.LONG_TYPE));

    //@DependsOn(next::class) TODO
    public FieldIdentifier previous = fieldIdentifier(fieldWrapper -> fieldWrapper.isOfType(TypeUtilities.getTypeOfIdentifiedClass("Node")));
//    && fieldWrapper.getName().equals(TypeUtilities.getIdentifiedField("next"))); TODO

    public MethodIdentifier remove = methodIdentifier(methodWrapper -> methodWrapper.returnTypeEquals(Type.VOID_TYPE));

//    @DependsOn(hasNext::class) TODO
//    class next : UniqueMapper.InMethod.Field(hasNext::class) {
//        override val predicate = predicateOf<Instruction2> { it.opcode == Opcodes.GETFIELD }
//    }
}

File: find-names/src/main/java/mapper/identifiers/classes/Node.kt
package mapper.identifiers.classes

import mapper.abstractclasses.IdentityMapper
import mapper.abstractclasses.UniqueMapper
import mapper.annotations.DependsOn
import mapper.annotations.MethodParameters
import mapper.predicateutilities.and
import mapper.predicateutilities.predicateOf
import mapper.predicateutilities.type
import mapper.wrappers.ClassWrapper
import mapper.wrappers.FieldWrapper
import mapper.wrappers.InstructionMapper
import mapper.wrappers.MethodWrapper
import org.objectweb.asm.Opcodes
import org.objectweb.asm.Type.*

class Node : IdentityMapper.Class() {
    override val predicate = predicateOf<ClassWrapper> { it.instanceFields.size == 3 }
            .and { it.instanceFields.count { it.type == LONG_TYPE } == 1 }
            .and { it.superType == Any::class.type }
            .and { c -> c.instanceFields.count { it.type == c.type } == 2 }

    class key : IdentityMapper.InstanceField() {
        override val predicate = predicateOf<FieldWrapper> { it.type == LONG_TYPE }
    }

    @DependsOn(next::class)
    class previous : IdentityMapper.InstanceField() {
        override val predicate = predicateOf<FieldWrapper> { it.type == type<Node>() }
                .and { it != field<next>() }
    }

    @DependsOn(hasNext::class)
    class next : UniqueMapper.InMethod.Field(hasNext::class) {
        override val predicate = predicateOf<InstructionMapper> { it.opcode == Opcodes.GETFIELD }
    }

    @MethodParameters()
    class hasNext : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == BOOLEAN_TYPE }
    }

    @MethodParameters()
    class remove : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == VOID_TYPE }
    }
}